omegaconf/dictconfig.py

```diff
@@ -631,6 +631,10 @@ def _set_value_impl(
                     value,
                     allow_objects=self._get_flag("allow_objects"),
                 )
+                # Structured Configs that are a subclass of Dict gets their struct flag set to False by default.
+                if is_dict(get_type_of(value)):
+                    self._set_flag("struct", False)
```

Comment:
I'm struggling a bit to fully understand what this means. At least in part because I'm not familiar with this chunk of code.

Something that surprises me is that if `value` is a kind of dictionary, it seems like we lose the dictionary content in the assignment, because we use the content of `data`, and `data` does not contain what is in the dictionary. Here is an example:

```python
@dataclass
class Str2Str(Dict[str, str]):
    x: int = 0

a = Str2Str()
assert a.x == 0
a["y"] = 1
assert a["y"] == 1
assert get_structured_config_data(a) == {"x": 0}
```

Is that intended?

Reply:
Yes, this is worrisome (though not related to this PR specifically).
To finish the above example:
```python
cfg_a = OmegaConf.create(a)
assert cfg_a.x == 0
assert "y" not in cfg_a
```

I'd call that a bug.

Reply:
Yes. this looks like a bug. I am going to give some context about the struct flag, structured configs, Dict fields and types extending Dict in a top level comment.